Packages/CandidateUI/Sources/CandidateUI/VerticalCandidateController.swift

```diff
@@ -0,0 +1,560 @@
+        }
+
+        scrollView.documentView = tableView
+        panel.contentView?.addSubview(scrollView)
+
+        let paraStyle = NSMutableParagraphStyle()
+        paraStyle.setParagraphStyle(NSParagraphStyle.default)
+        paraStyle.firstLineHeadIndent = candidateTextLeftMargin
+        paraStyle.lineBreakMode = .byClipping
+
+        candidateTextParagraphStyle = paraStyle
+
+        super.init(window: panel)
+        tableView.dataSource = self
+        tableView.delegate = self
+        tableView.doubleAction = #selector(rowDoubleClicked(_:))
+        tableView.target = self
+
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(boundsChange),
+                                               name: NSView.boundsDidChangeNotification,
+                                               object: scrollView.contentView)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    public override func reloadData() {
+        maxCandidateAttrStringWidth = ceil(candidateFont.pointSize + candidateTextPadding)
+        tableView.reloadData()
+        layoutCandidateView()
+        if delegate?.candidateCountForController(self) ?? 0 > 0 {
+            selectedCandidateIndex = 0
+        }
+    }
+
+    public override func showNextPage() -> Bool {
+        scrollPageByOne(true)
+    }
+
+    public override func showPreviousPage() -> Bool {
+        scrollPageByOne(false)
+    }
+
+    public override func highlightNextCandidate() -> Bool {
+        moveSelectionByOne(true)
+    }
+
+    public override func highlightPreviousCandidate() -> Bool {
+        moveSelectionByOne(false)
+    }
+
+    public override func candidateIndexAtKeyLabelIndex(_ index: UInt) -> UInt {
+        guard let delegate = delegate else {
+            return UInt.max
+        }
+
+        let firstVisibleRow = tableView.row(at: scrollView.documentVisibleRect.origin)
+        if firstVisibleRow != -1 {
+            let result = UInt(firstVisibleRow) + index
+            if result < delegate.candidateCountForController(self) {
+                return result
+            }
+        }
+
+        return UInt.max
+    }
+
+    public override var selectedCandidateIndex: UInt {
+        get {
+            let selectedRow = tableView.selectedRow
+            return selectedRow == -1 ? UInt.max : UInt(selectedRow)
+
+        }
+        set {
+            guard let delegate = delegate else {
+                return
+            }
+            var newIndex = newValue
+            let selectedRow = tableView.selectedRow
+            let labelCount = keyLabels.count
+            let itemCount = delegate.candidateCountForController(self)
+
+            if newIndex == UInt.max {
+                if itemCount == 0 {
+                    tableView.deselectAll(self)
+                    return
+                }
+                newIndex = 0
```

Comment:
The return on line 285 should be put right after here. The check on line 284 is never true because `newIndex` is now always 0 if `newValue` is set to `UInt.max`, so the code still crashes...